lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java

```diff
@@ -257,4 +256,79 @@ public void writeDoesNotAllowJavaEnumForGenericEnum() throws IOException {
 
     writer.write(record, encoder);
   }
+
+  @Test
+  public void testExternalSchemaWithExplicitNullDefault() throws Exception {
+    Schema schema = createExternalSchemaWithExplicitNullDefault();
+    GenericRecord record = createRecord(schema);
+    writeObject(schema, record);
+  }
+
+  @Test
+  public void testExternalSchemaWithoutExplicitNullDefault() throws Exception {
+    Schema schema = createExternalSchemaWithoutExplicitNullDefault();
+    GenericRecord record = createRecord(schema);
+    writeObject(schema, record);
+  }
+
+  @Test
+  public void testCodeGeneratedSchemaWithoutExplicitNullDefault() throws Exception {
+    Schema schema = createCodeGeneratedSchemaWithoutExplicitNullDefault();
+    GenericRecord record = createRecord(schema);
+    writeObject(schema, record);
+  }
+
+  @Test
+  public void testCodeGeneratedSchemaWithExplicitNullDefault() throws Exception {
+    Schema schema = createCodeGeneratedSchemaWithExplicitNullDefault();
+    GenericRecord record = createRecord(schema);
+    writeObject(schema, record);
+  }
+
+  private Schema createExternalSchemaWithExplicitNullDefault() {
+    String s = "{\"type\":\"record\",\"name\":\"my_record\",\"namespace\":\"mytest.namespace\",\"doc\":\"doc\"," +
+            "\"fields\":[{\"name\":\"f\",\"type\":[\"null\",\"string\"],\"doc\":\"field doc doc\", " +
+            "\"default\":null}]}";
+    return createSchemaFromString(s);
+  }
+
+  private Schema createExternalSchemaWithoutExplicitNullDefault() {
+    String s = "{\"type\":\"record\",\"name\":\"my_record\",\"namespace\":\"mytest.namespace\",\"doc\":\"doc\"," +
+            "\"fields\":[{\"name\":\"f\",\"type\":[\"null\",\"string\"],\"doc\":\"field doc doc\"}]}";
+    return createSchemaFromString(s);
+  }
+
+  private Schema createSchemaFromString(String schema) {
+    return new Schema.Parser().parse(schema);
+  }
+
+  private Schema createCodeGeneratedSchemaWithoutExplicitNullDefault() {
+    Schema schema = Schema.createRecord("my_record", "doc", "mytest.namespace", false);
+    Schema.Field stringField = new Schema.Field("f",
+            Schema.createUnion(Schema.create(Schema.Type.NULL), Schema.create(Schema.Type.STRING)),
+            "field doc doc", null);
+    schema.setFields(Arrays.asList(stringField));
+    return schema;
+  }
+
+  private Schema createCodeGeneratedSchemaWithExplicitNullDefault() {
+    Schema schema = Schema.createRecord("my_record", "doc", "mytest.namespace", false);
+    Schema.Field stringField = new Schema.Field("f",
+            Schema.createUnion(Schema.create(Schema.Type.NULL), Schema.create(Schema.Type.STRING)),
+            "field doc doc", org.apache.avro.JsonProperties.NULL_VALUE);
+    schema.setFields(Arrays.asList(stringField));
+    return schema;
+  }
+
+  private void writeObject(Schema schema, GenericRecord datum) throws Exception {
+    BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(new ByteArrayOutputStream(5), null);
```

Comment:
Good point, I don't think there is a good reason for using 5. Removed the parameter, using the default now.